test/EFCore.PG.FunctionalTests/TestUtilities/NpgsqlDatabaseCleaner.cs

```diff
@@ -93,6 +63,82 @@ public override void Clean(DatabaseFacade facade)
             base.Clean(facade);
         }
 
+        void DropExtensions(NpgsqlConnection conn)
+        {
+            const string getExtensions = @"
+SELECT name FROM pg_available_extensions WHERE installed_version IS NOT NULL AND name <> 'plpgsql'";
+
+            List<string> extensions;
+            using (var cmd = new NpgsqlCommand(getExtensions, conn))
+            {
+                using var reader = cmd.ExecuteReader();
+                extensions = reader.Cast<DbDataRecord>().Select(r => r.GetString(0)).ToList();
+            }
+
+            if (extensions.Any())
+            {
+                var dropExtensionsSql = string.Join("", extensions.Select(e => $"DROP EXTENSION \"{e}\" CASCADE;"));
+                using var cmd = new NpgsqlCommand(dropExtensionsSql, conn);
+                cmd.ExecuteNonQuery();
+            }
+        }
+
+        /// <summary>
+        /// Drop user-defined ranges and enums, cascading to all tables which depend on them
+        /// </summary>
+        void DropTypes(NpgsqlConnection conn)
+        {
+            const string getUserDefinedRangesEnums = @"
+SELECT ns.nspname, typname
+FROM pg_type
+JOIN pg_namespace AS ns ON ns.oid = pg_type.typnamespace
+WHERE typtype IN ('r', 'e') AND nspname <> 'pg_catalog'";
+
+            List<(string Schema, string Name)> userDefinedTypes;
+            using (var cmd = new NpgsqlCommand(getUserDefinedRangesEnums, conn))
+            {
+                using var reader = cmd.ExecuteReader();
+                userDefinedTypes = reader.Cast<DbDataRecord>().Select(r => (r.GetString(0), r.GetString(1))).ToList();
+            }
+
+            if (userDefinedTypes.Any())
+            {
+                var dropTypes = string.Join("", userDefinedTypes.Select(t => $@"DROP TYPE ""{t.Schema}"".""{t.Name}"" CASCADE;"));
```

Comment:
`Join` with an empty string as separator behaves the same way as `Concat`. And you don't need this complex block.

```csharp

using var cmd = new NpgsqlCommand(getUserDefinedRangesEnums, conn);
using var reader = cmd.ExecuteReader();
var dropTypes = string.Concat(reader
    .Cast<DbDataRecord>()
    .Select(r => $@"DROP TYPE ""{r.GetString(0)}"".""{r.GetString(1)}"" CASCADE;"));

if (dropTypes.Length > 0)
{
    cmd.CommandText = dropTypes;
    cmd.ExecuteNonQuery();
}
```

Reply:
Switched from Join to Concat, thanks.

However, the previous reader must be closed/disposed before the second command can be executed...